contrib/slurm/slurmexecutor.py

```diff
@@ -251,57 +283,120 @@ def run_slurm(benchmark, args, log_file):
+        def get_checked_seff_result():
+            seff_result = subprocess.run(
+                seff_command,
+                stdout=subprocess.PIPE,
+                stderr=subprocess.STDOUT,
+            )
+            if "exit code" in str(seff_result.stdout):
+                return seff_result
+            else:
+                return None
 
-    return {
-        "walltime": wall_time,
-        "cputime": cpu_time,
-        "memory": memory_usage,
-        "exitcode": ProcessExitCode.create(value=exit_code),
-    }
+        # sometimes `seff` needs a few extra seconds to realize the task has ended
+        result = wait_for(get_checked_seff_result, 30, 2)
 
+        status, exit_code, cpu_time, wall_time, memory_usage = parse_seff(
+            str(result.stdout)
+        )
 
-exit_code_pattern = re.compile(r"exit code (\d+)")
+        if os.path.exists(exitcode_file):
+            with open(exitcode_file, "r") as f:
+                returncode = int(f.read())
+                logging.debug("Exit code in file %s: %d", exitcode_file, returncode)
+        else:
+            if status == "COMPLETED":
+                raise Exception(
+                    "Should never happen: exit code not found, but task was reported COMPLETED."
+                )
+            logging.debug("Exit code not found in file: %s", exitcode_file)
+            returncode = 0
+
+        ret = {
+            "walltime": wall_time,
+            "cputime": cpu_time,
+            "memory": memory_usage,
+            "exitcode": ProcessExitCode.create(value=returncode),
+        }
+
+        if status != "COMPLETED":
+            ret["terminationreason"] = {
+                "OUT_OF_MEMORY": "memory",
+                "OUT_OF_ME+": "memory",
+                "TIMEOUT": "cputime",
+                "ERROR": "failed",
+                "FAILED": "killed",
+                "CANCELLED": "killed",
+            }.get(status, status)
+
+            if ret["terminationreason"] == "memory":
+                ret["memory"] = memory
+            if ret["terminationreason"] == "cputime":
+                ret["cputime"] = timelimit
+
+        # Runexec would populate the first 6 lines with metadata
+        with open(log_file, "w+") as file:
+            with open(tmp_log, "r") as log_source:
+                content = log_source.read()
+                file.write(f"jobid: {jobid}\n")
+                file.write(
+                    f"srun output (retcode: {str(returncode)}), truncated: {str(first_lines)}\n"
+                )
+                file.write(f"seff output: {str(result.stdout)}\n")
+                file.write(f"Parsed data: {str(ret)}\n")
+                empty_lines = "\n" * 3
+                file.write(empty_lines)
+                file.write("\n")
+                file.write(content)
+                if content == "":
+                    file.write("Original log file did not contain anything.")
```

Comment:
If I understand correctly, this is for writing the log file that will be part of the BenchExec results?

This should look the same as if you run BenchExec locally. In particular, the first line should have the executed command line. This is *extremely* useful because you can just copy the whole line for executing the same thing locally outside of BenchExec.

All the other information here seems like debugging information that is useful for the developers of the SLURM mode. But we don't want debugging information in the tool log. If you really want to store that in a file, then it should be a separate file that is only written with `--debug`.

Reply:
I think you're right. I will make this behave more like the rest of benchexec. What should I put in the other 5 lines? Unfortunately the first 6 lines have to be padded, otherwise it's cut off before the toolinfo modules get a chance of parsing the output: https://github.com/sosy-lab/benchexec/blob/a75bb4f9745e7a605818f1fa5ffcdef147990772/benchexec/model.py#L1094
(Also, I thought based on this comment, that putting some useful logs here would be OK, since that's what these 6 lines are for -- maybe this comment should be removed/rephrased, as to not confuse further developers?)

Reply:
After the line with the output, there should be the same as from standard execution:
```python
                "\n\n\n"
                + "-" * 80
                + "\n\n\n"
```
Can we keep the debug output separate? It is already not so nice (but there since forever and really useful) that we have the command line in the same file as the tool output. But for post-processing, it is really better to keep different things separate. For example, debug output could get misinterpreted by the tool-info module. And if you run in debug mode, you might still want to put the results somewhere for other people to see, so let's not confuse them with debug output.

Reply:
The tool-info module will never get this result, as the first 6 lines get cut off by the code in `module.py` before the tool-info modules get a chance of looking at it. But if you think it's better to keep it separate, I'm open to suggestions on where to put that (maybe a different file altogether)? If I create a new file next to the logfile, will it be properly handled by later parts of the code, e.g., when zipping the logfiles?

Reply:
Yes, a separate file next to the log file with some additional suffix (`run.log_file + ".something"`) is what [similar modules](https://github.com/sosy-lab/benchexec/blob/main/contrib/vcloud/benchmarkclient_executor.py) also use. These files won't be zipped then.